src/tokens/ERC1155.vy

```diff
@@ -111,65 +180,88 @@ def supportsInterface(interface_id: bytes4) -> bool:
+        acceptance magic value.
     @param owner The 20-byte address which previously
-           owned the token.
+        owned the token.
     @param to The 20-byte receiver address.
     @param id The 32-byte identifier of the token.
     @param amount The 32-byte token amount that is
-           being transferred.
+        being transferred.
     @param data The maximum 1024-byte additional data
-           with no specified format.
+        with no specified format.
     """
-    pass
+    assert owner == msg.sender or self.isApprovedForAll[owner][msg.sender], "ERC1155: Not authorized to transfer"
+    assert to != empty(address), "ERC1155: transfer to the zero address"
+    assert self._balances[owner][id] >= amount, "ERC1155: insufficient balance"
+
+    self._balances[owner][id] = unsafe_sub(self._balances[owner][id], amount)
+    self._balances[to][id] = unsafe_add(self._balances[to][id], amount)
+
+    log TransferSingle(msg.sender, owner, to, id, amount)
+    assert self._check_on_erc1155_received(owner, to, id, amount, data), "ERC1155: transfer to non-ERC1155Receiver implementer"
 
 
 @external
-def safeBatchTransferFrom(owner: address, to: address, ids: DynArray[uint256, max_value(uint16)], amounts: DynArray[uint256, max_value(uint16)],
-                          data: Bytes[1024]):
+def safeBatchTransferFrom(
+    owner: address,
+    to: address,
+    ids: DynArray[uint256, max_value(uint16)],
+    amounts: DynArray[uint256, max_value(uint16)],
+    data: Bytes[1024]
+):
     """
     @dev Batched version of `safeTransferFrom`.
     @notice Note that `ids` and `amounts` must have the
-            same length. Also, if `to` refers to a smart
-            contract, it must implement {IERC1155Receiver-onERC1155Received}
-            and return the acceptance magic value.
+        same length. Also, if `to` refers to a smart
+        contract, it must implement {IERC1155Receiver-onERC1155Received}
+        and return the acceptance magic value.
     @param owner The 20-byte address which previously
-           owned the token.
+        owned the token.
     @param to The 20-byte receiver address.
     @param ids The 32-byte array of token identifiers. Note
-           that the order and length must match the 32-byte
-           `amounts` array.
+        that the order and length must match the 32-byte
+        `amounts` array.
     @param amounts The 32-byte array of token amounts that are being
-           transferred. Note that the order and length must match
-           the 32-byte `ids` array.
+        transferred. Note that the order and length must match
+        the 32-byte `ids` array.
     @param data The maximum 1024-byte additional data
-           with no specified format.
+        with no specified format.
     """
-    pass
+    assert owner == msg.sender or self.isApprovedForAll[owner][msg.sender], "ERC1155: Not authorized to transfer"
+    assert to != empty(address), "ERC1155: transfer to the zero address"
```

Comment:
everything that comes here could be put into an internal function `_safe_batch_transfer_from`. See reasoning above.